Add arrow-adbc packages

lidavidm opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Announced by the Apache Arrow project here: https://lists.apache.org/thread/lkplzc80gkghdjt30zrtp4spyrmbdpcz

ADBC is an API for database access for the Apache Arrow ecosystem. This PR adds the 0.1.0 release, including these packages:
- A driver manager for C/C++ and Python to dynamically load ADBC drivers (analogously to ODBC, JDBC),
- Drivers for PostgreSQL and SQLite.

Future releases will add more functionality. 

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/arrow-adbc```) and found it was in an excellent condition.

lidavidm commented:
CC @mariusvniekerk, how does this look? Probably I'll have to tweak it once it lands to get the arm64 builds working

lidavidm commented:
@conda-forge/help-python-c, ready for review!

jorisvandenbossche commented on recipes/arrow-adbc/meta.yaml:
```diff
@@ -0,0 +1,203 @@
+        - cross-python_{{ target_platform }}                                     # [build_platform != target_platform]
+        - cython
+        - pip
+        - poetry
```
poetry is not used anymore nowadays and can be removed?

lidavidm replied:
Thanks for catching that. Removed.

carterbox replied:
Poetry is still listed here...

carterbox commented on recipes/arrow-adbc/meta.yaml:
```diff
@@ -0,0 +1,203 @@
+  number: 0
+
+outputs:
+  - name: adbc-driver-manager-cpp
```
Outputs with headers and shared libraries should have run_exports which pin an ABI compatible version of the output.

carterbox commented on recipes/arrow-adbc/build-python.sh:
```diff
@@ -0,0 +1,29 @@
+export SETUPTOOLS_SCM_PRETEND_VERSION=$PKG_VERSION
+
+echo "==== INSTALL ${PKG_NAME}"
+$PYTHON -m pip install . -vvv --no-deps
```
```suggestion
$PYTHON -m pip install . -vvv --no-deps --no-build-isolation
```

carterbox commented on recipes/arrow-adbc/build-python.bat:
```diff
@@ -0,0 +1,25 @@
+set SETUPTOOLS_SCM_PRETEND_VERSION=%PKG_VERSION%
+
+echo "==== INSTALL %PKG_NAME%"
+%PYTHON% -m pip install . -vvv --no-deps || exit /B 1
```
```suggestion
%PYTHON% -m pip install . -vvv --no-deps --no-build-isolation || exit /B 1
```

carterbox commented on recipes/arrow-adbc/meta.yaml:
```diff
@@ -0,0 +1,203 @@
+        - test -d $PREFIX/lib/cmake/AdbcDriverManager/                           # [unix]
+        - test -f $PREFIX/lib/pkgconfig/adbc-driver-manager.pc                   # [unix]
+        - test ! -f $PREFIX/lib/libadbc_driver_manager.a                         # [unix]
+        - test -f $PREFIX/lib/libadbc_driver_manager.so                          # [linux]
+        - test -f $PREFIX/lib/libadbc_driver_manager.dylib                       # [osx]
```
I suggest checking for a specific major SO name in order to guard against ABI breakages.

carterbox commented on recipes/arrow-adbc/meta.yaml:
```diff
@@ -0,0 +1,203 @@
+    requirements:
+      build:
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
```
There are linker warnings that `libstdcxx` is unused by the sqlite library. Either remove this compiler or add `ignore_run_exports_from: {{ compiler('cxx') }}`.

carterbox commented on recipes/arrow-adbc/meta.yaml:
```diff
@@ -0,0 +1,203 @@
+    script: build-python.bat                                                     # [win]
+    requirements:
+      build:
+        # We don't use these, but Conda needs them to know that this is a binary package.
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
```
This is OK as long as these shared objects are being built here with our build tools and not elsewhere.

carterbox commented on recipes/arrow-adbc/meta.yaml:
```diff
@@ -0,0 +1,203 @@
+        - if not exist %LIBRARY_LIB%\cmake\AdbcDriverManager exit 1              # [win]
+        - if not exist %LIBRARY_LIB%\pkgconfig\adbc-driver-manager.pc exit 1     # [win]
+
+  - name: adbc-driver-postgresql-cpp
```
The name of these cpp packages should be `libadbc-driver-postgresql`

carterbox commented:
I still need to review the actual build artifacts, but here's some preliminary feedback.

carterbox commented on recipes/arrow-adbc/meta.yaml:
```diff
@@ -0,0 +1,203 @@
+        - if not exist %LIBRARY_LIB%\cmake\AdbcDriverSQLite exit 1               # [win]
+        - if not exist %LIBRARY_LIB%\pkgconfig\adbc-driver-sqlite.pc exit 1      # [win]
+
+  - name: adbc-driver-manager
```
adbc-driver-manager is shipping headers and c source code which are not needed

carterbox commented on recipes/arrow-adbc/build-python.sh:
```diff
@@ -0,0 +1,29 @@
+export SETUPTOOLS_SCM_PRETEND_VERSION=$PKG_VERSION
+
+echo "==== INSTALL ${PKG_NAME}"
+$PYTHON -m pip install . -vvv --no-deps --no-build-isolation
```
The python packages are being built multiple times for python=3.11 only. Although the build matrix claims they are being built for various 3.8, 3.9, 3.10, 3.11. Not sure why that is.

lidavidm commented:
Thanks for the review, I'll fix these when I get a chance.

Note that we're coming up on 0.2.0 which has a number of improvements which I'll sync back here (in particular, the Python packages can become noarch and the workaround isn't needed).

lidavidm commented:
I've punted on macOS and Windows since a bunch of linking issues came up that I can't replicate on VMs/M1. So for now this package will have to be Linux only.

- Synced with upstream recipe (though now this recipe is ahead of upstream)
- Renamed packages
- Removed source/headers from Python adbc-driver-manager
- Added run-exports
- Updated checks to test for specific soname

I'm not sure what's going on with Python. One thing I was considering was one feedstock for packages with native libraries and one for the pure-Python bits that depend on those native libraries.

jorisvandenbossche commented:
I suppose it should be possible to avoid running the MacOS/Windows CI at all? (currently they take almost the same time as linux to set up some build envs, but still only ~10min, so not a big deal)

Otherwise looks good based on my limited conda-forge experience.

lidavidm commented:
Once there's a feedstock, we can disable osx/windows in conda-forge.yml
